advanced/custom_integration/impl.py

```diff
@@ -0,0 +1,113 @@
+import requests
+from requests.exceptions import RequestException
+from sym.sdk.annotations import hook, reducer
+from sym.sdk.integrations import pagerduty, slack
+from sym.sdk.templates import ApprovalTemplate
+
+
+# Reducers fill in the blanks that your workflow needs in order to run.
+@reducer
+def get_approvers(event):
+    """Route Sym requests to a channel specified in the sym_flow."""
+
+    # allow_self lets the requester approve themself, which is great for testing!
+    return slack.channel("#sym-requests", allow_self=True)
+
+
+# Hooks let you customize workflow behavior running code before or after each
+# step in your workflow.
+@hook
+def on_request(event):
+    """If the requester is on-call, auto-approve their requests"""
+    try:
+        if is_requester_on_call(event):
+            original_reason = event.payload.fields["reason"]
+            return ApprovalTemplate.approve(
+                reason=f"Auto-approved On-call engineer: {original_reason}️"
+            )
+    except (RuntimeError, RequestException) as e:
+        # Catch any exceptions, such as an expired API Token.
+        # Skip auto-approval and report the error to the error channel.
+        print(
+            f"An error occurred while checking who's on-call. Falling back to off-call behavior.\n{e}"
+        )
+
+
+def api_request(api_path, integration, params={}):
+    """
+    Hit the given path in the VictorOps API using our custom integration.
+    For details on the API format:
+    https://portal.victorops.com/public/api-docs.html
+    """
+
+    # Get the api key we passed in to the integration with the secret_ids_json setting
+    token = integration.settings["secrets"][0].retrieve_value()
+
+    headers = {
+        "Accept": "application/json",
+        # We used the API ID as our custom integration's external ID
+        "X-VO-Api-Id": integration.external_id,
+        "X-VO-Api-Key": token,
+    }
+    response = requests.get(
+        f"https://api.victorops.com/api-public/{api_path}",
+        headers=headers,
+        params=params,
+    )
+    body = response.json()
+
+    if not response.ok:
+        message = body.get("message", "")
+        raise RuntimeError(f"API failed with message: {message}")
+
+    return body
+
+
+def find_user_by_email(email, integration):
+    """Hit the VictorOps API to find the username for the given email"""
+    response = api_request("v2/user", integration, params={"email": email})
+    for user in response.get("users", []):
+        return user["username"]
+
+
+def get_custom_user(user, integration):
+    """
+    Each Sym user may have a separate identity stored for each integrated service.
+    Get the VictorOps username associated with the Sym user.
```

Comment:
Nit: I'd opt for some more detailed comments? Maybe something like
```suggestion
    Get the VictorOps username associated with the Sym user.
    If the user already has a VictorOps identity, then return it. Otherwise, fetch the identity from VictorOps and persist it.
```